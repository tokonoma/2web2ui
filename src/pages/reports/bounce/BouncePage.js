import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import _ from 'lodash';
import { safeRate } from 'src/helpers/math';
import { refreshBounceReport } from 'src/actions/bounceReport';
import { TableCollection, Empty, LongTextContainer } from 'src/components';
import { Percent } from 'src/components/formatters';
import { PanelLoading } from 'src/components/loading';
import { Page, Tabs } from 'src/components/matchbox';
import ReportOptions from '../components/ReportOptions';
import BounceChart from './components/BounceChart';
import MetricsSummary from '../components/MetricsSummary';
import { mapStateToProps } from 'src/selectors/bounceReport';
import AddFilterLink from '../components/AddFilterLink';

const columns = [
  { label: 'Reason', width: '45%', sortKey: 'reason' },
  { label: 'Domain', sortKey: 'domain' },
  { label: 'Category', sortKey: 'bounce_category_name' },
  { label: 'Classification', sortKey: 'classification_id' },
  { label: 'Count (%)', sortKey: 'count_bounce' },
];

export class BouncePage extends Component {
  state = {
    tab: 0,
  };

  componentDidUpdate(prevProps) {
    if (prevProps.reportOptions !== this.props.reportOptions) {
      this.props.refreshBounceReport(this.props.reportOptions);
    }
  }

  getRowData = item => {
    const { aggregates = {} } = this.props;
    const {
      reason,
      domain,
      bounce_category_name,
      bounce_class_name,
      count_bounce,
      count_admin_bounce,
    } = item;
    // calculate the rate of admin bounces against all bounces
    let numerator = count_bounce;
    let denominator = aggregates.countBounce;

    if (bounce_category_name === 'Admin') {
      numerator = count_admin_bounce;
      denominator += aggregates.countAdminBounce;
    }

    return [
      <LongTextContainer text={reason} />,
      <AddFilterLink
        newFilter={{ type: 'Recipient Domain', value: domain }}
        reportType={'bounce'}
        content={domain}
      />,
      bounce_category_name,
      bounce_class_name,
      <span>
        {numerator}&nbsp;
        <small>
          (<Percent value={safeRate(numerator, denominator)} />)
        </small>
      </span>,
    ];
  };

  renderChart() {
    const { chartLoading, aggregates, categories, types, adminCategories } = this.props;

    if (!chartLoading) {
      if (
        _.isEmpty(aggregates) ||
        (!aggregates.countAdminBounce && this.state.tab === 1) ||
        (!aggregates.countBounce && this.state.tab === 0)
      ) {
        return <Empty message="No bounces to report" />;
      }
    }

    return (
      <BounceChart
        loading={chartLoading}
        aggregates={aggregates}
        categories={categories}
        types={types}
        admin={adminCategories}
        tab={this.state.tab}
      />
    );
  }

  renderCollection() {
    const { tableLoading } = this.props;
    const reasons = this.state.tab === 1 ? this.props.adminReasons : this.props.reasons;

    if (tableLoading) {
      return <PanelLoading />;
    }

    if (!reasons.length) {
      return <Empty message={'No bounce reasons to report'} />;
    }

    return (
      <TableCollection
        columns={columns}
        rows={reasons}
        getRowData={this.getRowData}
        defaultSortColumn="count_bounce"
        defaultSortDirection="desc"
        pagination
        filterBox={{
          show: true,
          keyMap: {
            category: 'bounce_category_name',
            classification: 'bounce_class_name',
          },
          itemToStringKeys: ['bounce_category_name', 'bounce_class_name', 'domain', 'reason'],
          exampleModifiers: ['domain', 'category', 'classification'],
          matchThreshold: 5,
        }}
      />
    );
  }

  renderTopLevelMetrics() {
    const { chartLoading, aggregates } = this.props;
    const { countBounce, countSent, countAdminBounce } = aggregates;

    // TODO Add support doc link - <UnstyledLink to={LINKS.ADMIN_BOUNCE} external>Learn more</UnstyledLink>.
    const adminBounceText = countAdminBounce ? (
      <Fragment>
        {countAdminBounce.toLocaleString()} messages were categorized as Admin Bounces.
      </Fragment>
    ) : null;

    // Aggregates aren't ready until chart refreshes
    if (chartLoading) {
      return <PanelLoading minHeight="115px" />;
    }

    if (_.isEmpty(aggregates)) {
      return null;
    }

    return (
      <MetricsSummary
        rateValue={safeRate(countBounce, countSent)}
        rateTitle="Bounce Rate"
        secondaryMessage={adminBounceText}
      >
        <strong>{countBounce.toLocaleString()}</strong> of{' '}
        <strong>{countSent.toLocaleString()}</strong> sent messages bounced
      </MetricsSummary>
    );
  }

  renderTabs() {
    return (
      <Tabs
        tabs={[
          { content: 'Bounces', onClick: () => this.handleTab(0) },
          { content: 'Admin Bounces', onClick: () => this.handleTab(1) },
        ]}
        selected={this.state.tab}
      />
    );
  }

  handleTab = index => {
    this.setState({ tab: index });
  };

  render() {
    const { chartLoading, bounceSearchOptions } = this.props;
    return (
      <Page title="Bounce Report">
        <ReportOptions reportLoading={chartLoading} searchOptions={bounceSearchOptions} />
        {this.renderTopLevelMetrics()}
        {this.renderTabs()}
        {this.renderChart()}
        <hr />
        {this.renderCollection()}
      </Page>
    );
  }
}

const mapDispatchToProps = {
  refreshBounceReport,
};

export default connect(mapStateToProps, mapDispatchToProps)(BouncePage);
