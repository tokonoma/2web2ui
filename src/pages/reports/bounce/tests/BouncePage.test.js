import React from 'react';
import { BouncePage } from '../BouncePage';
import { shallow } from 'enzyme';
import MetricsSummary from '../../components/MetricsSummary';

describe('BouncePage: ', () => {
  const props = {
    loading: false,
    aggregates: {
      countBounce: 1,
      countSent: 10,
      countAdminBounce: 5,
    },
    reportOptions: {
      relativeRange: 'hour',
    },
    addFilters: jest.fn(),
    refreshBounceReport: jest.fn(),
    reasons: [
      {
        bounce_category_name: 'Block',
        bounce_class_description:
          'The message was blocked by the receiver as coming from a known spam source',
        bounce_class_name: 'Spam Block',
        classification_id: 51,
        count_bounce: 5,
        domain: 'yahoo.com',
        reason:
          "554 - 5.7.1 Blocklisted by 'twoomail.com'(twoo.com.multi.surbl.org) Contact the postmaster of this domain for resolution. This attempt has been logged.",
      },
    ],
    adminReasons: [
      {
        bounce_class_name: 'Admin Failure',
        bounce_class_description: 'The message was failed by eceleritys configured policies',
        bounce_category_name: 'Admin',
        classification_id: 25,
        count_admin_bounce: 2,
        domain: 'yahoo.com',
        reason: '550 - Administrative prohibition #51',
      },
    ],
    bounceSearchOptions: {},
  };

  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<BouncePage {...props} />);
  });

  it('should render', () => {
    expect(props.refreshBounceReport).not.toHaveBeenCalled();
    expect(wrapper).toMatchSnapshot();
  });

  it('should refresh when report options reference changes', () => {
    const newReportOptions = {};
    expect(props.refreshBounceReport).not.toHaveBeenCalled();
    wrapper.setProps({ reportOptions: newReportOptions });
    expect(props.refreshBounceReport).toHaveBeenCalledWith(newReportOptions);
  });

  it('should render correctly with no bounces', () => {
    wrapper.setProps({ aggregates: {} });
    expect(wrapper).toMatchSnapshot();
  });

  it('should show loading panel when table is loading', () => {
    wrapper.setProps({ tableLoading: true });
    expect(wrapper).toMatchSnapshot();
  });

  it('should display loading panel on top level metrics when aggregates are loading', () => {
    wrapper.setProps({ chartLoading: true });
    expect(wrapper).toMatchSnapshot();
  });

  it('should not display top level metrics admin bounce message with no admin bounces', () => {
    wrapper.setProps({
      aggregates: {
        countBounce: 1,
        countSent: 10,
        countAdminBounce: 0,
      },
    });
    expect(wrapper.find(MetricsSummary).props().secondaryMessage).toBe(null);
  });

  it('should not display top level metrics when there are no aggregates', () => {
    wrapper.setProps({ aggregates: {} });
    expect(wrapper.find(MetricsSummary)).toHaveLength(0);
  });

  it('should show empty reasons table when there are no reasons when on regular bounce tab', () => {
    wrapper.setState({ tab: 0 });
    wrapper.setProps({
      reasons: [],
      adminReasons: [
        {
          bounce_class_name: 'Admin Failure',
          bounce_class_description: 'The message was failed by eceleritys configured policies',
          bounce_category_name: 'Admin',
          classification_id: 25,
          count_admin_bounce: 2,
          domain: 'yahoo.com',
          reason: '550 - Administrative prohibition #51',
        },
      ],
    });
    expect(wrapper).toMatchSnapshot();
  });

  it('should show empty reasons table when there are no admin reasons when on admin bounce tab', () => {
    wrapper.setState({ tab: 1 });
    wrapper.setProps({
      adminReasons: [],
      reasons: [
        {
          bounce_category_name: 'Block',
          bounce_class_description:
            'The message was blocked by the receiver as coming from a known spam source',
          bounce_class_name: 'Spam Block',
          classification_id: 51,
          count_bounce: 5,
          domain: 'yahoo.com',
          reason:
            "554 - 5.7.1 Blocklisted by 'twoomail.com'(twoo.com.multi.surbl.org) Contact the postmaster of this domain for resolution. This attempt has been logged.",
        },
      ],
    });
    expect(wrapper).toMatchSnapshot();
  });

  it('should render row data properly', () => {
    const rows = wrapper.instance().getRowData({
      reason: 'u y bounce?',
      domain: 'gmail',
      bounce_category_name: 'bouncy',
      count_bounce: 10,
    });

    expect(rows).toMatchSnapshot();
  });

  it('should render row data for admin bounce properly', () => {
    const rows = wrapper.instance().getRowData({
      reason: 'bouncy boi',
      domain: 'gmail',
      bounce_category_name: 'Admin',
      count_admin_bounce: 5,
    });

    expect(rows).toMatchSnapshot();
  });

  it('should change state when the tab is clicked', () => {
    wrapper.instance().handleTab(1);
    expect(wrapper).toHaveState('tab', 1);
  });
});
