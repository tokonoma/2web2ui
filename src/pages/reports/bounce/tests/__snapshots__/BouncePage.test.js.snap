// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BouncePage:  should display loading panel on top level metrics when aggregates are loading 1`] = `
<Page
  title="Bounce Report"
>
  <withRouter(Connect(ReportOptions))
    reportLoading={true}
    searchOptions={Object {}}
  />
  <PanelLoading
    accent={false}
    minHeight="115px"
  />
  <Panel>
    <TabsWrapper>
      <Tabs
        selected={0}
        tabs={
          Array [
            Object {
              "content": "Bounces",
              "onClick": [Function],
            },
            Object {
              "content": "Admin Bounces",
              "onClick": [Function],
            },
          ]
        }
      />
    </TabsWrapper>
    <Panel.Section>
      <BounceChart
        aggregates={
          Object {
            "countAdminBounce": 5,
            "countBounce": 1,
            "countSent": 10,
          }
        }
        loading={true}
        tab={0}
      />
    </Panel.Section>
  </Panel>
  <hr />
  <TableCollection
    columns={
      Array [
        Object {
          "label": "Reason",
          "sortKey": "reason",
          "width": "45%",
        },
        Object {
          "label": "Domain",
          "sortKey": "domain",
        },
        Object {
          "label": "Category",
          "sortKey": "bounce_category_name",
        },
        Object {
          "label": "Classification",
          "sortKey": "classification_id",
        },
        Object {
          "label": "Count (%)",
          "sortKey": "count_bounce",
        },
      ]
    }
    defaultSortColumn="count_bounce"
    defaultSortDirection="desc"
    filterBox={
      Object {
        "exampleModifiers": Array [
          "domain",
          "category",
          "classification",
        ],
        "itemToStringKeys": Array [
          "bounce_category_name",
          "bounce_class_name",
          "domain",
          "reason",
        ],
        "keyMap": Object {
          "category": "bounce_category_name",
          "classification": "bounce_class_name",
        },
        "matchThreshold": 5,
        "show": true,
      }
    }
    getRowData={[Function]}
    pagination={true}
    rows={
      Array [
        Object {
          "bounce_category_name": "Block",
          "bounce_class_description": "The message was blocked by the receiver as coming from a known spam source",
          "bounce_class_name": "Spam Block",
          "classification_id": 51,
          "count_bounce": 5,
          "domain": "yahoo.com",
          "reason": "554 - 5.7.1 Blocklisted by 'twoomail.com'(twoo.com.multi.surbl.org) Contact the postmaster of this domain for resolution. This attempt has been logged.",
        },
      ]
    }
  />
</Page>
`;

exports[`BouncePage:  should render 1`] = `
<Page
  title="Bounce Report"
>
  <withRouter(Connect(ReportOptions))
    searchOptions={Object {}}
  />
  <Connect(MetricsSummary)
    rateTitle="Bounce Rate"
    rateValue={10}
    secondaryMessage={
      <React.Fragment>
        5
         messages were categorized as Admin Bounces.
      </React.Fragment>
    }
  >
    <strong>
      1
    </strong>
     of
     
    <strong>
      10
    </strong>
     sent messages bounced
  </Connect(MetricsSummary)>
  <Panel>
    <TabsWrapper>
      <Tabs
        selected={0}
        tabs={
          Array [
            Object {
              "content": "Bounces",
              "onClick": [Function],
            },
            Object {
              "content": "Admin Bounces",
              "onClick": [Function],
            },
          ]
        }
      />
    </TabsWrapper>
    <Panel.Section>
      <BounceChart
        aggregates={
          Object {
            "countAdminBounce": 5,
            "countBounce": 1,
            "countSent": 10,
          }
        }
        tab={0}
      />
    </Panel.Section>
  </Panel>
  <hr />
  <TableCollection
    columns={
      Array [
        Object {
          "label": "Reason",
          "sortKey": "reason",
          "width": "45%",
        },
        Object {
          "label": "Domain",
          "sortKey": "domain",
        },
        Object {
          "label": "Category",
          "sortKey": "bounce_category_name",
        },
        Object {
          "label": "Classification",
          "sortKey": "classification_id",
        },
        Object {
          "label": "Count (%)",
          "sortKey": "count_bounce",
        },
      ]
    }
    defaultSortColumn="count_bounce"
    defaultSortDirection="desc"
    filterBox={
      Object {
        "exampleModifiers": Array [
          "domain",
          "category",
          "classification",
        ],
        "itemToStringKeys": Array [
          "bounce_category_name",
          "bounce_class_name",
          "domain",
          "reason",
        ],
        "keyMap": Object {
          "category": "bounce_category_name",
          "classification": "bounce_class_name",
        },
        "matchThreshold": 5,
        "show": true,
      }
    }
    getRowData={[Function]}
    pagination={true}
    rows={
      Array [
        Object {
          "bounce_category_name": "Block",
          "bounce_class_description": "The message was blocked by the receiver as coming from a known spam source",
          "bounce_class_name": "Spam Block",
          "classification_id": 51,
          "count_bounce": 5,
          "domain": "yahoo.com",
          "reason": "554 - 5.7.1 Blocklisted by 'twoomail.com'(twoo.com.multi.surbl.org) Contact the postmaster of this domain for resolution. This attempt has been logged.",
        },
      ]
    }
  />
</Page>
`;

exports[`BouncePage:  should render correctly with no bounces 1`] = `
<Page
  title="Bounce Report"
>
  <withRouter(Connect(ReportOptions))
    searchOptions={Object {}}
  />
  <Panel>
    <TabsWrapper>
      <Tabs
        selected={0}
        tabs={
          Array [
            Object {
              "content": "Bounces",
              "onClick": [Function],
            },
            Object {
              "content": "Admin Bounces",
              "onClick": [Function],
            },
          ]
        }
      />
    </TabsWrapper>
    <Empty
      message="No bounces to report"
    />
  </Panel>
  <hr />
  <TableCollection
    columns={
      Array [
        Object {
          "label": "Reason",
          "sortKey": "reason",
          "width": "45%",
        },
        Object {
          "label": "Domain",
          "sortKey": "domain",
        },
        Object {
          "label": "Category",
          "sortKey": "bounce_category_name",
        },
        Object {
          "label": "Classification",
          "sortKey": "classification_id",
        },
        Object {
          "label": "Count (%)",
          "sortKey": "count_bounce",
        },
      ]
    }
    defaultSortColumn="count_bounce"
    defaultSortDirection="desc"
    filterBox={
      Object {
        "exampleModifiers": Array [
          "domain",
          "category",
          "classification",
        ],
        "itemToStringKeys": Array [
          "bounce_category_name",
          "bounce_class_name",
          "domain",
          "reason",
        ],
        "keyMap": Object {
          "category": "bounce_category_name",
          "classification": "bounce_class_name",
        },
        "matchThreshold": 5,
        "show": true,
      }
    }
    getRowData={[Function]}
    pagination={true}
    rows={
      Array [
        Object {
          "bounce_category_name": "Block",
          "bounce_class_description": "The message was blocked by the receiver as coming from a known spam source",
          "bounce_class_name": "Spam Block",
          "classification_id": 51,
          "count_bounce": 5,
          "domain": "yahoo.com",
          "reason": "554 - 5.7.1 Blocklisted by 'twoomail.com'(twoo.com.multi.surbl.org) Contact the postmaster of this domain for resolution. This attempt has been logged.",
        },
      ]
    }
  />
</Page>
`;

exports[`BouncePage:  should render row data for admin bounce properly 1`] = `
Array [
  <LongTextContainer
    text="bouncy boi"
  />,
  <Connect(AddFilterLink)
    content="gmail"
    newFilter={
      Object {
        "type": "Recipient Domain",
        "value": "gmail",
      }
    }
    reportType="bounce"
  />,
  "Admin",
  undefined,
  <span>
    5
     
    <small>
      (
      <Percent
        value={83.33333333333334}
      />
      )
    </small>
  </span>,
]
`;

exports[`BouncePage:  should render row data properly 1`] = `
Array [
  <LongTextContainer
    text="u y bounce?"
  />,
  <Connect(AddFilterLink)
    content="gmail"
    newFilter={
      Object {
        "type": "Recipient Domain",
        "value": "gmail",
      }
    }
    reportType="bounce"
  />,
  "bouncy",
  undefined,
  <span>
    10
     
    <small>
      (
      <Percent
        value={1000}
      />
      )
    </small>
  </span>,
]
`;

exports[`BouncePage:  should show empty reasons table when there are no admin reasons when on admin bounce tab 1`] = `
<Page
  title="Bounce Report"
>
  <withRouter(Connect(ReportOptions))
    searchOptions={Object {}}
  />
  <Connect(MetricsSummary)
    rateTitle="Bounce Rate"
    rateValue={10}
    secondaryMessage={
      <React.Fragment>
        5
         messages were categorized as Admin Bounces.
      </React.Fragment>
    }
  >
    <strong>
      1
    </strong>
     of
     
    <strong>
      10
    </strong>
     sent messages bounced
  </Connect(MetricsSummary)>
  <Panel>
    <TabsWrapper>
      <Tabs
        selected={1}
        tabs={
          Array [
            Object {
              "content": "Bounces",
              "onClick": [Function],
            },
            Object {
              "content": "Admin Bounces",
              "onClick": [Function],
            },
          ]
        }
      />
    </TabsWrapper>
    <Panel.Section>
      <BounceChart
        aggregates={
          Object {
            "countAdminBounce": 5,
            "countBounce": 1,
            "countSent": 10,
          }
        }
        tab={1}
      />
    </Panel.Section>
  </Panel>
  <hr />
  <Panel>
    <Empty
      message="No bounce reasons to report"
    />
  </Panel>
</Page>
`;

exports[`BouncePage:  should show empty reasons table when there are no reasons when on regular bounce tab 1`] = `
<Page
  title="Bounce Report"
>
  <withRouter(Connect(ReportOptions))
    searchOptions={Object {}}
  />
  <Connect(MetricsSummary)
    rateTitle="Bounce Rate"
    rateValue={10}
    secondaryMessage={
      <React.Fragment>
        5
         messages were categorized as Admin Bounces.
      </React.Fragment>
    }
  >
    <strong>
      1
    </strong>
     of
     
    <strong>
      10
    </strong>
     sent messages bounced
  </Connect(MetricsSummary)>
  <Panel>
    <TabsWrapper>
      <Tabs
        selected={0}
        tabs={
          Array [
            Object {
              "content": "Bounces",
              "onClick": [Function],
            },
            Object {
              "content": "Admin Bounces",
              "onClick": [Function],
            },
          ]
        }
      />
    </TabsWrapper>
    <Panel.Section>
      <BounceChart
        aggregates={
          Object {
            "countAdminBounce": 5,
            "countBounce": 1,
            "countSent": 10,
          }
        }
        tab={0}
      />
    </Panel.Section>
  </Panel>
  <hr />
  <Panel>
    <Empty
      message="No bounce reasons to report"
    />
  </Panel>
</Page>
`;

exports[`BouncePage:  should show loading panel when table is loading 1`] = `
<Page
  title="Bounce Report"
>
  <withRouter(Connect(ReportOptions))
    searchOptions={Object {}}
  />
  <Connect(MetricsSummary)
    rateTitle="Bounce Rate"
    rateValue={10}
    secondaryMessage={
      <React.Fragment>
        5
         messages were categorized as Admin Bounces.
      </React.Fragment>
    }
  >
    <strong>
      1
    </strong>
     of
     
    <strong>
      10
    </strong>
     sent messages bounced
  </Connect(MetricsSummary)>
  <Panel>
    <TabsWrapper>
      <Tabs
        selected={0}
        tabs={
          Array [
            Object {
              "content": "Bounces",
              "onClick": [Function],
            },
            Object {
              "content": "Admin Bounces",
              "onClick": [Function],
            },
          ]
        }
      />
    </TabsWrapper>
    <Panel.Section>
      <BounceChart
        aggregates={
          Object {
            "countAdminBounce": 5,
            "countBounce": 1,
            "countSent": 10,
          }
        }
        tab={0}
      />
    </Panel.Section>
  </Panel>
  <hr />
  <PanelLoading
    accent={false}
    minHeight="400px"
  />
</Page>
`;
